Improve deprecation_warning.h and fix some missing / improper deprecated macros

MaelRL opened the pull request:
## Summary of Changes

[This commit](https://github.com/CGAL/cgal/commit/6d13fa2c3178dfecd30b04bcefc5861304466ce8) and other commits deprecated Regular_triangulation_euclidean_traits, but the macros `CGAL_NO_DEPRECATED_CODE` are missing in the files.

## Release Management

* Affected package(s): Triangulation_2/3
* Issue(s) solved (if any): --
* Feature/Small Feature (if any): --

MaelRL committed:
Added CGAL_NO_DEPRECATED_CODE macros to Regular_triangulation_euclidean_traits

janetournois commented on Triangulation_2/include/CGAL/Regular_triangulation_euclidean_traits_2.h:
```diff
@@ -24,6 +24,11 @@
 
 #include <CGAL/license/Triangulation_2.h>
 
+#ifdef CGAL_NO_DEPRECATED_CODE
```
I think this should be `#ifndef`, instead of `#ifdef`
And, should it really make an error? A warning would be enough, maybe...?

sloriot commented:
See also `CGAL/git/cgal/Installation/include/CGAL/internal/deprecation_warning.h`

janetournois commented:
thank you!

lrineau commented:
The macro `CGAL_NO_DEPRECATED_CODE` is supposed to trigger errors if deprecated code is used.

http://doc.cgal.org/4.9/Manual/preliminaries.html:
> In order to help users manage the changes to apply to their code, we attempt to make CGAL code emit warnings when deprecated code is used. This can be done using some compiler specific features. Those warnings can be disabled by defining the macro CGAL_NO_DEPRECATION_WARNINGS. On top of this, we also provide a macro, CGAL_NO_DEPRECATED_CODE, which, when defined, disables all deprecated features. This allows users to easily test if their code relies on deprecated features.

MaelRL commented:
Was successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-8.shtml

lrineau commented:
As pointed out by me and Sébastien in https://github.com/CGAL/cgal/pull/2054#issuecomment-294152430 and https://github.com/CGAL/cgal/pull/2054#issuecomment-294156350, deprecated headers should:
  - error out if `CGAL_NO_DEPRECATED_CODE` is defined, 
  - display a warning, unless `CGAL_NO_DEPRECATION_WARNINGS` is defined.

Both is already handled by the use of `<CGAL/internal/deprecation_warning.h>`. See for example `Git/cgal/AABB_tree/include/CGAL/AABB_polyhedron_triangle_primitive.h` for a usage.

MaelRL commented:
I disregarded those comments because `deprecation_warning.h` does not allow to pass custom messages and there is no replacement header for these deprecated classes. Also because there are other CGAL classes that are deprecated and do not use the `deprecation_warning.h` include.

Two possibilities:
1. I use `deprecation_warning.h` and the user must go look up in the code/doc what's wrong.
2. I change the code of `deprecation_warning.h` such that a macro like `CGAL_INTERNAL_DEPRECATED_MESSAGE_DETAILS` is appended (if it exists) to `CGAL_INTERNAL_DEPRECATED_MESSAGE`.

lrineau commented:
I would prefer the second choice (improvement of `deprecation_warning`). Would you mind amending also the corresponding documentation in our guidelines? Currently is says:
> To deprecate something, she should make modifications in the code and in the documentation.
>  * In the .h code-file she wants to deprecate, she should add the following :
>
>     #define CGAL_DEPRECATED_HEADER "<CGAL/old_header_file.h>"
>     #define CGAL_REPLACEMENT_HEADER "<CGAL/new_header_file.h>"
>     #include <CGAL/internal/deprecation_warning.h>


https://github.com/CGAL/cgal/wiki/Developing-Features-with-Git#coding

MaelRL committed:
Improved deprecation_warning.h

-- Actually fail compilation even if CGAL_NO_DEPRECATION_WARNINGS is defined
-- An optional message can be passed through CGAL_DEPRECATED_MESSAGE_DETAILS
-- Print the replacement header in all cases, not just when warning about
   the deprecated header.

MaelRL commented:
If the name `CGAL_DEPRECATED_MESSAGE_DETAILS` for the optional custom text macro is deemed alright, I will edit the doc.

I made a few additional changes, see 6550621 commit message.

MaelRL committed:
Fixed improper guards against the inclusion of deprecated headers

MaelRL committed:
Removed some warnings

MaelRL commented:
d3bb2c2 adds a lot of warnings due to two deprecated headers that are often included:
- `Regular_triangulation_euclidean_traits_3.h`. All of these warnings are already fixed in the implicit conversion branch, which is almost ready to be merged, so I'd rather not do anything about the warnings in this branch.
- `Weighted_point.h`. `Surface_mesher` did not get converted to the kernel's weighted points and the class `Weighted_point` is used a lot. I don't really have time to look into it, so do I just comment the include `<CGAL/internal/depreciation_warning.h>` for now ?

MaelRL committed:
Partial revert of c3f32e612978fd6d0e3d8cd65963c886217b003e

Surface_mesher currently needs io_signatures for the deprecated Weighted_point

MaelRL committed:
Silenced depreciation warnings from Regular_triangulation_euclidean_traits_23

MaelRL committed:
Silenced depreciation warnings from Weighted_point.h

MaelRL commented:
I silenced both deprecations so this branch is clean. I'll re-enable the depreciation macros of `Regular_triangulation_euclidean_traits_23` in the implicit conversion branch (thus this branch must be merged first). For `Weighted_point.h`... someone will need to clean `Surface_mesher` some day.

If the `new deprecation_warning.h` is satisfying, this is ready to be tested.

afabri committed:
Change comment to avoid warning

afabri committed:
Change comment to avoid warning

MaelRL committed:
Change comment to avoid warning

MaelRL committed:
Temporarily silenced weighted alpha shape traits depreciation warnings

MaelRL commented:
Rebased and all warnings fixed.

MaelRL commented:
I changed the base to master because it's not urgent stuff and I think it's better if it's in the same release as the branch that removes implicit conversions between P/WP (which will re-enable the temporarily warnings that were temporarily silenced here).

lrineau committed:
Re-add <CGAL/config.h> to avoid a conflict with `master`

The `#include` was removed by error by:

    73122ac202e1d24b9d31537c4872ed9ba9dbd71d
    Author: Andreas Fabri <andreas.fabri@geometryfactory.com>
    Date:   Tue May 16 08:26:08 2017 +0200

        Change comment to avoid warning

lrineau commented on Installation/include/CGAL/internal/deprecation_warning.h:
```diff
@@ -12,44 +12,69 @@
-// don't trigger on NO_DEPRECATIOON_WARNINGS and don't trigger twice on NO_DEPRECATED_CODE
-#if !defined(CGAL_NO_DEPRECATION_WARNINGS) && !defined(CGAL_NO_DEPRECATED_CODE)
+#if defined(CGAL_NO_DEPRECATED_CODE) // No deprecated code.
+CGAL_static_assertion_msg(false, CGAL_INTERNAL_DEPRECATED_MESSAGE);
```
@MaelRL:
(This in an old PR, from about 10 years ago.)

Why not a plain `#error`, instead of a static assertion with `false` as the condition?
